// @flmngr/flmngr-server-node-express/dist/mjs/ExpressRequest.js
export class ExpressRequest {
    constructor(req) {
        this.req = req;
    }
    getParameterNumber(name, defaultValue) {
        let strValue = this.getParameterString(name, null);
        let isCorrectNumber = true;
        if (strValue !== null) {
            isCorrectNumber = "" + parseInt(strValue) === strValue;
        }
        else {
            isCorrectNumber = false;
        }
        if (isCorrectNumber)
            return parseInt(strValue);
        else
            return defaultValue;
    }
    getParameterString(name, defaultValue) {
        let value = this.req.body[name];
        if (typeof value === "string")
            return value;
        else
            return defaultValue;
    }
    getParameterStringArray(name, defaultValue) {
        let value = this.req.body[name];
        if (typeof value === "string")
            value = [value];
        if (Array.isArray(value))
            return value;
        else
            return defaultValue;
    }
    getParameterFile(name) {
        if (!("postFile" in this.req))
            return null;
        return {
            data: this.req["postFile"]["data"],
            fileName: this.req["postFile"]["filename"]["filename"]
        };
    }
}
