// @flmngr/flmngr-server-node/dist/cjs/model/Message.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Message = void 0;
class Message {
    constructor(isCacheIssue) {
        this.isCacheIssue = isCacheIssue;
    }
    static createMessage(isCacheException, code, arg1 = null, arg2 = null, arg3 = null, parentException = null) {
        let msg = new Message(isCacheException);
        msg.code = code;
        msg.parentException = parentException;
        if (arg1 != null) {
            msg.args = [];
            msg.args.push(arg1);
            if (arg2 != null) {
                msg.args.push(arg2);
                if (arg3 != null) {
                    msg.args.push(arg3);
                }
            }
        }
        return msg;
    }
    getJSON() {
        return {
            code: this.code,
            args: this.args
        };
    }
    getParentException() {
        return this.parentException;
    }
}
exports.Message = Message;
Message.FILE_ERROR_SYNTAX = -1;
Message.FILE_ERROR_DOES_NOT_EXIST = -2;
Message.FILE_ERROR_INCORRECT_IMAGE_EXT_CHANGE = -3;
Message.ACTION_NOT_FOUND = 0;
Message.UNABLE_TO_CREATE_UPLOAD_DIR = 1;
Message.UPLOAD_ID_NOT_SET = 2;
Message.UPLOAD_ID_INCORRECT = 3;
Message.MALFORMED_REQUEST = 4;
Message.NO_FILE_UPLOADED = 5;
Message.FILE_SIZE_EXCEEDS_LIMIT = 6;
Message.INCORRECT_EXTENSION = 7;
Message.WRITING_FILE_ERROR = 8;
Message.UNABLE_TO_DELETE_UPLOAD_DIR = 9;
Message.UNABLE_TO_DELETE_FILE = 10;
Message.DIR_DOES_NOT_EXIST = 11;
Message.FILES_NOT_SET = 12;
Message.FILE_IS_NOT_IMAGE = 13;
Message.DUPLICATE_NAME = 14;
Message.FILE_ALREADY_EXISTS = 15;
Message.FILES_ERRORS = 16;
Message.UNABLE_TO_COPY_FILE = 17;
Message.IMAGE_PROCESS_ERROR = 18;
Message.MAX_RESIZE_WIDTH_EXCEEDED = 19;
Message.MAX_RESIZE_HEIGHT_EXCEEDED = 20;
Message.UNABLE_TO_WRITE_IMAGE_TO_FILE = 21;
Message.INTERNAL_ERROR = 22;
Message.DOWNLOAD_FAIL_CODE = 23;
Message.DOWNLOAD_FAIL_IO = 24;
Message.DOWNLOAD_FAIL_HOST_DENIED = 25;
Message.DOWNLOAD_FAIL_INCORRECT_URL = 26;
Message.FILE_SIZE_EXCEEDS_SYSTEM_LIMIT = 29;
Message.FILE_SIZE_EXCEEDS_SYSTEM_LIMIT_2 = 30;
Message.FM_FILE_DOES_NOT_EXIST = 10001;
Message.FM_UNABLE_TO_WRITE_PREVIEW_IN_CACHE_DIR = 10002;
Message.FM_UNABLE_TO_CREATE_PREVIEW = 10003;
Message.FM_DIR_NAME_CONTAINS_INVALID_SYMBOLS = 10004;
Message.FM_DIR_NAME_INCORRECT_ROOT = 10005;
Message.FM_FILE_IS_NOT_IMAGE = 10006;
Message.FM_ROOT_DIR_DOES_NOT_EXIST = 10007;
Message.FM_UNABLE_TO_LIST_CHILDREN_IN_DIRECTORY = 10008;
Message.FM_UNABLE_TO_DELETE_DIRECTORY = 10009;
Message.FM_UNABLE_TO_CREATE_DIRECTORY = 10010;
Message.FM_UNABLE_TO_RENAME = 10011;
Message.FM_DIR_CANNOT_BE_READ = 10012;
Message.FM_ERROR_ON_COPYING_FILES = 10013;
Message.FM_ERROR_ON_MOVING_FILES = 10014;
Message.FM_NOT_ERROR_NOT_NEEDED_TO_UPDATE = 10015;
Message.FM_ROOT_DIR_IS_NOT_SET = 10016;
Message.FM_DIR_IS_NOT_READABLE = 10017;
Message.FM_DIR_IS_NOT_WRITABLE = 10018;
